### ISSUE TITLE
NameError: name 'shortBAM' is not defined

### ISSUE THREAD

**Issue by peterthorpe5** [nextgenusfs/funannotate#266]
Dear Funannotate,

firstly. Love this tool well done!!!
I masked, then trained, the predicted my genes and it crashed on "update" in the section where it calls stringtie. (  File "/home/linuxbrew/funannotate/bin/funannotate-update.py", line 1866, in <module>
    if lib.which('stringtie') and lib.checkannotations(shortBAM):
NameError: name 'shortBAM' is not defined)..... could you please share you thoughts on how to fix this error?

cheers,
Pete

train="funannotate train -i ${genome} -o ${out_folder} --left ${left}  --right ${right} \
	--trinity Trinity-GG.fasta --species ${species} --cpus ${threads} \
	--strain ${strain} --max_intronlen ${max_intron_len}"
echo ${train}
eval ${train}

predict="funannotate predict -i ${genome} -o ${out_folder} \
    --species  ${species} --strain ${strain} --cpus ${threads} --max_intronlen ${max_intron_len} --organism other \
	--busco_db nematoda --database /home/linuxbrew/data/databases/\
	--organism other --augustus_species GPAL_newton_softmasked \
	--transcript_alignments Gpal_newtontest03_newton.valid_blat_alignments.gff3 \
	--rna_bam Gp_soft_maskedAligned.sortedByCoord.out.bam "
echo ${predict} 
eval ${predict}
	

**Are you using the latest release?**
v1.5.0 docker image

**Describe the bug**
~/data/funannotate$ funannotate update -i fun_newton --cpus 6
Traceback (most recent call last):
  File "/home/linuxbrew/funannotate/bin/funannotate-update.py", line 1866, in <module>
    if lib.which('stringtie') and lib.checkannotations(shortBAM):
NameError: name 'shortBAM' is not defined

**What command did you issue?**
.funannotate update -i fun_newton --cpus 6

**Logfiles**
Please provide relavent log files of the error. 

**OS/Install Information**
 -  output of `funannotate check --show-versions`

Checking dependencies for funannotate v1.5.0
-------------------------------------------------------
You are running Python v 2.7.14. Now checking python packages...
biopython: 1.68
goatools: 0.8.4
matplotlib: 2.2.2
natsort: 5.3.3
numpy: 1.15.0
pandas: 0.23.4
psutil: 5.4.6
requests: 2.19.1
scikit-learn: 0.19.2
scipy: 1.1.0
seaborn: 0.9.0
All 11 python packages installed

You are running Perl v 5.026002. Now checking perl modules...
Bio::Perl: 1.007002
Carp: 1.38
Clone: 0.39
DBD::SQLite: 1.58
DBD::mysql: 4.046
DBI: 1.641
DB_File: 1.835
Data::Dumper: 2.167
File::Basename: 2.85
File::Which: 1.20
Getopt::Long: 2.5
Hash::Merge: 0.300
JSON: 2.97001
LWP::UserAgent: 6.35
Logger::Simple: 2.0
POSIX: 1.76
Parallel::ForkManager: 1.20
Pod::Usage: 1.69
Scalar::Util::Numeric: 0.40
Storable: 3.11
Text::Soundex: 3.05
Thread::Queue: 3.13
Tie::File: 1.02
URI::Escape: 3.31
YAML: 1.26
threads: 2.21
threads::shared: 1.58
All 27 Perl modules installed

**Comment by nextgenusfs** [comment]
Hi @peterthorpe5, glad you are finding it useful.  Thanks for reporting.  I think this might have been addressed here b854ac200003aef8bce74979c76355a6540fd9ed or is at least a similar issue, basically that variable was not being properly defined in some use cases.  Looks like you are using docker, so when next release gets cut it will be fixed, hopefully.  In the meantime, you might be able to pass the reads again to `funannotate update` and have it remap to the genome with `hisat2` -- I know not ideal and will see if I can't get a new release tagged in the next few days.

**Comment by peterthorpe5** [comment]
Thank you very much for your reply.

FYI: I also had an issue with Diamond search with know proteins (returns 0 after filtering). When funannotate copies over the sequences to a fasta file a sneaky not ASCI character get imbedded in proteins.combined.fa. If this is simply reformatted as a fasta with biopython, diamond can make a database with it.

Cheers,
Pete

**Comment by peterthorpe5** [comment]
Docker feature request: Could you please include an editor in the image (I never fixed internet access issue from within the image). So Vi or Nano wold be useful .

Cheers,

Pete

**Comment by nextgenusfs** [comment]
added `nano` to next docker-base build. https://github.com/nextgenusfs/funannotate/commit/29d23d0fc2f4325e84b3ee8f5d7059886a37dbfc

**Comment by nextgenusfs** [comment]
Per the fasta headers -- there is actually a routine that I try to clean up the different FASTA headers for the protein input here: https://github.com/nextgenusfs/funannotate/blob/29d23d0fc2f4325e84b3ee8f5d7059886a37dbfc/lib/library.py#L1162. So it is using the SeqIO biopython module actually, but apparently strange characters still get through. If you have some suggestions on how to improve that would be great.

**Comment by peterthorpe5** [comment]
Thank you for adding nano to the docker image.

As for the protein file. I couldn’t get this step to run during the predict stage of the software, so I went through the code (found the relevant part) and ran Diamond by itself which complained about a non ASCII character. I couldn’t find this character when searching through the file. But if I reformatted it by doing the following. Once it was reformatted it worked I am afraid since this file is fixed I cant reproduce the error):

from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio import SeqIO

def reformat_as_fasta(filename, outfile):
    "this function re-write a file as a fasta file. Input = fasta file. Output = fasta out file."
    f= open(outfile, 'w')
    for seq_record in SeqIO.parse(filename, "fasta"):
        seq_record.seq = seq_record.seq.upper()
        SeqIO.write(seq_record, f, "fasta")
    f.close()

cheers,

Pete
